### ISSUE TITLE
LibTorch crashes in debug mode

### ISSUE THREAD

**Issue by tholzmann** [pytorch/pytorch#22681]
## 🐛 Bug

When doing inference using LibTorch in C++ for semantic segmentation on the GPU, i get the following error in debug compilation mode:

cuda_dispatch_ptr INTERNAL ASSERT FAILED at C:\w\1\s\windows\pytorch\aten\src\ATen/native/DispatchStub.h:70, please report a bug to PyTorch. DispatchStub: missing CUDA kernel (operator () at C:\w\1\s\windows\pytorch\aten\src\ATen/native/DispatchStub.h:70)
(no backtrace available)

When I run it in release mode it does not crash and delivers the expected result. Also on the CPU, this assert is not shown.

## Environment

 - PyTorch Version (e.g., 1.0): latest nightly
 - OS (e.g., Linux): WIndows 10, Visual Studio 19
 - How you installed PyTorch (`conda`, `pip`, source): downloaded prebuild libtorch
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version: the one included in the prebuild (cuda 10)
 - GPU models and configuration: Titan X
 - Any other relevant information:

**Comment by peterjc123** [comment]
For MSVC, the debug/release configurations are not abi-compatible, which means you need to use the debug version of the library when you choose the debug configuration.

**Comment by peterjc123** [comment]
You may notice there are two download links for libtorch for Windows on pytorch.org and each of them is supposed to be used when you choose a specified configuration.

**Comment by tholzmann** [comment]
I checked it again now, I am using the correct debug/release version.

**Comment by ezyang** [comment]
This is a CUDA call right? Can you actually tell us what exactly the operator that's failing is? This sounds like some sort of static initializer problem. How are you linking with the CUDA version of the library? Basically, based on the error message, it seems that you are trying to call an operator, and we don't think that the CUDA version of this operator exists at all.

**Comment by tholzmann** [comment]
Yes, it is a cuda call. It already happens when trying to copy a tensor to the CUDA device by calling

inData = inData.to( *m_TorchDevice );

where inData is a torch::Tensor and m_TorchDevice is constructed by: 
m_TorchDevice = std::unique_ptr<torch::Device>(new torch::Device( torch::DeviceType::CUDA, 0 ))

**Comment by xsacha** [comment]
I am getting the same issue with latest nightly.
Only happens when using the Debug nightly.

**Comment by xsacha** [comment]
Still happening with 1st August 2019 nightly (Windows CUDA) if it helps.

**Comment by xsacha** [comment]
Still happening with 13th August 2019 nightly (Windows CUDA)

**Comment by ezyang** [comment]
Thanks for the update. We are a bit understaffed in the Windows fixes department, so this probably won't get fixed in the near future until someone who is using Windows rolls up their sleeves and figures out what is going on. I can help review a patch you submit.

**Comment by rje** [comment]
I've hit this as well and have started to dig in, but I'm fairly new to pytorch and this is my first time in the C++ codebase.

My test code is:

    #include <iostream>
    #include <torch/torch.h>
    #include <torch/script.h>

    int main() {
        auto hasCuda = torch::cuda::is_available();
        auto hasCudnn = torch::cuda::cudnn_is_available();
        std::cout << "Has cuda? " << hasCuda << " - Has cudnn? " << hasCudnn << std::endl;

        torch::jit::script::Module module;
        try {
            module = torch::jit::load("../../resnet18.pt");
            std::cout << "Ok!" << std::endl;
        }
        catch(c10::Error& error) {
            std::cout << "Error loading model: " << error.msg() << std::endl;
            return -1;
        }

        return 0;
    }

I started by putting the following debug line in front of the assert in DispatchStub.h:
`std::cout << "operator() for " << typeid(this).name() << " cuda dispatch ptr " << cuda_dispatch_ptr << " cpu dispatch ptr " << cpu_dispatch_ptr << " hip dispatch ptr " << hip_dispatch_ptr << std::endl;`

And when I run the application again I get the following result:
`operator() for struct at::native::DispatchStub<void (__cdecl*)(struct at::TensorIterator & __ptr64,bool),struct at::native::copy_stub> * __ptr64 cuda dispatch ptr 0000000000000000 cpu dispatch ptr 0000000000000000 hip dispatch ptr 0000000000000000`

So the next thing I checked was to see if RegisterCUDADispatch was being properly called for at::native::copy_stub, so I added the following debug line to RegisterCUDADispatch: 
`std::cout << "register cuda dispatch " << typeid(stub).name() << " - " << value << std::endl;`

I see 35 different results for this debug line when I start my test application, and sure enough a line for copy_stub is in there:
`register cuda dispatch struct at::native::DispatchStub<void (__cdecl*)(struct at::TensorIterator & __ptr64,bool),struct at::native::copy_stub> - 00007FFBCAF7AF30`

So it appears that the register call is happening correctly, but for some reason that value isn't being retained?

I'll continue to investigate, but open to suggestions if anyone has something they'd like me to try.

**Comment by peterjc123** [comment]
@rje Thanks for your investigation. Would you please post the stacktrace?

**Comment by ezyang** [comment]
> So it appears that the register call is happening correctly, but for some reason that value isn't being retained?

Could it be that there are two copies of the registry? Pasting out some addresses might help distinguish this situation.

**Comment by rje** [comment]
I went ahead and printed out the stub addresses. For copy_stub, we get the following:
    
    DispatchStub::RegisterCUDADispatch: Stub Type: struct at::native::DispatchStub<void (__cdecl*)(struct at::TensorIterator & __ptr64,bool),struct at::native::copy_stub>
        Stub Addr: 00007FF9F168B890

    DispatchStub::operator(): Stub Type: struct at::native::DispatchStub<void (__cdecl*)(struct at::TensorIterator & __ptr64,bool),struct at::native::copy_stub> * __ptr64
        Stub Addr: 00007FF9F168B890

So they seem like they should be the same object?

I'm going to keep annotating different aspects of the dispatch system to try and figure out what's going on, but very open to suggestion for things to try.

@peterjc123 Is there an easy way to get a stacktrace for libtorch on windows? All of the c10::Error messages have "(no backtrace available)"

**Comment by rje** [comment]
On a hunch that maybe initialization order was causing an issue I changed the following line in DispatchStub.h:

`FnPtr cuda_dispatch_ptr = nullptr;`

to:

`FnPtr cuda_dispatch_ptr;`

And after a rebuild, everything worked as expected.

I have some concerns about this as a 'fix' because I don't have a complete understanding of why that assignment is causing issues. Regardless, if it's acceptable to folks here I'm happy to submit a pull request with the change.

**Comment by peterjc123** [comment]
The internal error seems to be a VC bug.
https://stackoverflow.com/questions/53196700/inline-variable-is-initialized-more-than-once

**Comment by ezyang** [comment]
Wow, nice sleuthing!

**Comment by ezyang** [comment]
@peterjc123 Are you sure? It doesn't seem like the struct is declared inline here

**Comment by beichen2012** [comment]
I got the same error in : net->to(device) where the device is cuda:0

**Comment by ezyang** [comment]
@beichen2012 There is a partial (but unsound) patch which resolves this error in #24360 you might give it a try.

**Comment by beichen2012** [comment]
> @beichen2012 There is a partial (but unsound) patch which resolves this error in #24360 you might give it a try.

Thanks , but it only occurs in debug mode. so I will debug in cpu and publish in gpu...

**Comment by peterjc123** [comment]
Should be fixed by https://github.com/pytorch/pytorch/pull/24360.
